Feature/color on win

Cielquan opened the pull request:
## Problem:

The problem is the missing support for colors on windows machines and is described in #55.

## Changes: 

- I refactored the `devtools.debug.Debug._env_bool` classmethod as a function `env_bool` in `devtools.prettier`
  *Reason:* The method was not necessary to be a classmethod so I refactored it near `env_true` function which is used by the method.
- I added tests for new function `env_bool`.
- I introduced the `highlight` module which takes the roll to distinguish if highlighting is used. *More info below*
- I added tests for the new highlight`module.
- I changed `Debug.__init__` and `Debug.__call__` for the use of the new `highlight.use_highlight` function.

### `highlight` module

On import the module will, when on a windows machine, try to activate windows ANSI support by setting the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` flag via `SetConsoleMode`. See here for more info: https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences. Depending on the outcome the `color_active` variable will be set.

The code which does the work was posted here: https://bugs.python.org/msg291732. Truth be told I only kinda get what the code does but I think the author knows what he does. I tested it on two Win10 machines and it works.

Next there is the `use_highlight` function. It took over the work to distinguish if 
`DebugOutput.str` function called by `Debug.__call__` should use highlighting. The function will first check if the `Debug` class instance got created with `highlight=True`. If not set it will check the `PY_DEVTOOLS_HIGHLIGHT` environment variable. If also not set it decides by checking if `isatty(file_)` **and** `color_active` variable are `True`. The latter part is new and the variable gets set on import (see above). The part before is just the old behavior.

fixes #55 

**EDIT:** Info: My added tests succeed on windows but 21 others do not.

Cielquan commented:
I don't quite get the coverage thing but will look into it.

samuelcolvin commented on tests/test_main.py:
```diff
@@ -5,8 +5,10 @@
 
 import pytest
 
+import devtools.highlight
```
looks like you have this imported twice with different names.

Cielquan replied:
The marked import is for the monkeypatch and the other for direct usage. If you want I can remove the other and update the usage.

Cielquan commented on devtools/prettier.py:
```diff
@@ -42,6 +42,13 @@ def env_true(var_name, alt=None):
         return alt
 
 
+def env_bool(value, env_name, env_default):
```
I will look into it

Cielquan replied:
added type hints to `env_true` **and** `env_bool` because `env_bool` uses `env_true`

samuelcolvin commented on devtools/highlight.py:
```diff
@@ -0,0 +1,77 @@
+from .ansi import isatty
+from .prettier import env_bool
+
+if sys.platform == "win32":  # pragma: no cover (windows)
```
Maybe just mark them as xfail for now.

I hoped this pr would did many of them.

Cielquan replied:
- I replaced my double quotes with single quotes
- I added xfail to all (on windows) failing tests
- I added `windows` to `os`

Cielquan replied:
Unfortunately the CI test dont seem to run again since they all passed once ?!

samuelcolvin commented on devtools/highlight.py:
```diff
@@ -0,0 +1,77 @@
+from .ansi import isatty
+from .prettier import env_bool
+
+if sys.platform == 'win32':  # pragma: no cover (windows)
```
```suggestion
if sys.platform == 'win32':
```

samuelcolvin replied:
can you wrap this method in a function rather than in this sys test.

e.g. something like:

```py
def use_ansi_colours(file_):
    if sys.platform == 'win32':
        return use_ansi_colours_windows(file_)
    else:
        return use_ansi_colours_unix(file_)

def use_ansi_colours_windows(file_):
    ... this code

def use_ansi_colours_unix(file_):
    ... is isatty code
```

Cielquan replied:
Sure i refactored the sys test. 

The suggested removal of the `pragma`:
I changed the `pragma` and introduced `covdefaults`. This package lets you use platform specific `pragma`.

samuelcolvin commented on devtools/highlight.py:
```diff
@@ -0,0 +1,77 @@
+import sys
```
could you move all this code into `utils.py`

Cielquan replied:
Sure .. will rename `highlight.py` to `util.py`

Cielquan replied:
Done

samuelcolvin commented on devtools/highlight.py:
```diff
@@ -0,0 +1,77 @@
+            return _set_conout_mode(mode, mask)
+        except WindowsError as e:
+            if e.winerror == ERROR_INVALID_PARAMETER:
+                raise NotImplementedError
```
is this the only way `NotImplementedError` can be raised, if so, maybe better to just `return False`?

Cielquan replied:
I took the snippet and build around it. 

But I think I could take your proposal to simplify the code a bit. Will check.

Cielquan replied:
Done

samuelcolvin commented on tests/test_expr_render.py:
```diff
@@ -11,6 +11,7 @@ def foobar(a, b, c):
     return a + b + c
 
 
+@pytest.mark.xfail(sys.platform == "win32", reason="yet unknown windows problem")
```
single quotes please.

Cielquan replied:
Sorry .. habit .. will fix

Cielquan replied:
Done

samuelcolvin commented:
hi @Cielquan I'm afraid you still have lots of conflicts - see the PR summary above - that's what I meant by move into utils, not just rename.

Cielquan commented:
> hi @Cielquan I'm afraid you still have lots of conflicts - see the PR summary above - that's what I meant by move into utils, not just rename.

Year I rebased the up-to-date master into the branch .. but I need to fix some things .. the util.py was not there on the older master.

samuelcolvin commented:
> Year I rebased the up-to-date master into the branch .. but I need to fix some things .. the util.py was not there on the older master.

Yes sorry about that, my fault.

Cielquan commented:
> > Year I rebased the up-to-date master into the branch .. but I need to fix some things .. the util.py was not there on the older master.
> 
> Yes sorry about that, my fault.

No problem. Thats normal to update `master` is it not?

I think it should now work. Lets see.

samuelcolvin commented:
yes, it's normal, but I try to always merge PRs in the order they were created so people don't have to fix conflicts caused by new PRs. However in this case, I was rushing to get some things fixed, thus merged newer PRs before this one.

Cielquan commented:
Oh we got a circular import problem with the merge of highlight.py into  utils.py.

I will refactor a bit to fix it.

Cielquan commented:
Ok I refactored `env_bool` & `env_true` from `prettier` into `utils`.

I also moved tests into a new `test_utils.py` file.

Cielquan commented:
@samuelcolvin 
The `test_print` and `test_print_subprocess` tests are failing everywhere. I will look into it.

The windows pipline fails because it cannot find `twine`.

Cielquan commented:
So I fixed linux/mac with removal of `covdefaults` package which I just introduced. Problem is that it seems it does not work correctly with coverage directly .. `coverage html` fails while pytest with pytest-cov works. 
Problem is the `fail_under` config. Because of this I removed the package and changed the platform specific pragma to `no cover` for the windows parts. This should be fixed later when the windows tests are fixed.

The `windows`+ `twine` problem is another topic.

Cielquan commented:
Sooooo .. finally done :fireworks: 

The windows problem was `pip install -U setuptools pip` in ` make install`. On windows pip is unable to update itself. You need to call pip with python to update it: `python -m pip install -U setuptools pip`

Cielquan commented:
@samuelcolvin Is it good to merge or do you need me to change something? :smiley_cat:

samuelcolvin commented on devtools/utils.py:
```diff
@@ -1,11 +1,100 @@
+    kernel32.SetConsoleMode.errcheck = _check_bool
+    kernel32.SetConsoleMode.argtypes = (wintypes.HANDLE, wintypes.DWORD)
+
+    def _set_conout_mode(new_mode, mask=0xFFFFFFFF):
```
is there a reason this is a function, not inline below?

Cielquan replied:
I think for readability reasons. Sure you can nest those two try/except blocks, but IMO this is more readable.

If you want me to I can refactor it.

samuelcolvin replied:
I guess my feeling is that the extra function is less readable and a little more confusion - normally nested functions are only used when:
1. you need to avoid duplicated code
2. you need closure to access variables used within the function

I read this and immediately ask "where else is it being called, oh why not?", but this is fine. I'll merge once tests pass.

samuelcolvin commented on devtools/utils.py:
```diff
@@ -1,11 +1,100 @@
+        finally:
+            os.close(fdout)
+
+    def _enable_vt_mode() -> bool:
```
same with this?

Cielquan replied:
Ok this one is fine to be refactored. I think its a relic of the snippets origin.

Cielquan replied:
changed it

samuelcolvin commented on devtools/utils.py:
```diff
@@ -1,11 +1,100 @@
+
+
+def activate_win_color() -> bool:  # pragma: no cover
+    """Activate ANSI support on windows consoles.
```
```suggestion
    """
    Activate ANSI support on windows consoles.
```

samuelcolvin commented on devtools/utils.py:
```diff
@@ -1,11 +1,100 @@
+    if highlight is not None:
+        return highlight
+
+    if sys.platform == "win32":  # pragma: no cover
```
```suggestion
    if sys.platform == 'win32':  # pragma: no cover
```

Cielquan commented:
see above

samuelcolvin commented:
Thanks so much for this, sorry it took so long to merge.

I'll make a new release soon.

Cielquan commented:
> Thanks so much for this, sorry it took so long to merge.
> 
> I'll make a new release soon.

When do you think you can release the new version? 😃

samuelcolvin commented:
sorry 🙏, been busy 🏃 .

Just need to fix #62, then I can release. Will try to do so today.

samuelcolvin commented:
done, sorry for the delay.

Cielquan commented:
> done, sorry for the delay.

No problem. Thanks a lot 😄
